Support for external XMP SDK via conan

piponazo opened the pull request:
In this PR I added the needed changes in the conan & CMake code to handle the new XMP SDK 2016 version as an external dependency.

I wrote a conan recipe for XMPSDK that is maintained in a different repository:

https://github.com/piponazo/conan-xmpsdk

Packages for Linux, Max OSX and Windows are automatically generated and pushed to the piponazo's bintray repository:

https://bintray.com/piponazo/piponazo/XmpSdk%3Apiponazo

Note that this is not the main bintray repository, but anyways it is public and anyone can get the recipe and binary packages from there. I do not think we could ever get this recipe in the **conan-center** main repository. The XMP SDK CMake code is a bit of a mess, and the compilation fails for many configurations.

Note that the CMake code is prepared to handle the usage of the existing XMP version in our repository (when **EXIV2_ENABLE_XMP** is enabled, this is done by default), and for the external XMP version too (when **EXIV2_ENABLE_EXTERNAL_XMP** is enabled). 

However, since this is a experimental work, EXIV2_ENABLE_EXTERNAL_XMP is disabled in the CI jobs, and we will keep using the version included in our repository. 

In #297 we discussed briefly about the scenario where distributions could have a package for the XMP SDK. I think none of the main Linux distributions have a package for that library, and I really think we should not care about that possible scenario right now. 

Note that once this code is merged into master I can proceed with the deletion of the autotools and CMake code as we discussed in the team meeting. I did not want to do it before, because the autotools code was able to handle the compilation and usage of the external XMP SDK, while CMake was not prepared for that use case yet.

piponazo committed:
Add CMake code to handle external XMP

Experiment: Try with testing package

piponazo committed:
AppVeyor: Remove explicit option EXIV2_ENABLE_XMP. Already ON by default

piponazo committed:
Format README-CONAN.md file

piponazo committed:
Add info about adding the piponazo's conan repository

piponazo committed:
conan: Do not bring Expat when xmp=True.

Since the external XMP SDK is already adding the expat sources to the XMPCore library, we do not
need the Expat dependency for Exiv2.

piponazo committed:
CMake: FATAL_ERROR when both XMP cmake variables are one.

EXIV2_ENABLE_XMP and EXIV2_ENABLE_EXTERNAL_XMP are mutually exclusive.

piponazo committed:
Doc: Add documentation about how to use the external XMP SDK with conan & cmake

piponazo committed:
CMake: Print linker flags in the configuration summary

clanmills commented:
This seems wonderful to me.  I have to go out today, so I'll build and look at this tonight.

clanmills commented:
Can't get this build on MacOS-X.

```
827 rmills@rmillsmm:~/gnu/github/piponazo/exiv2/build $ python /Users/piponazo/envs/conan/lib/python2.7/site-packages/conans/conan.py install ..
....
with_libpsl=False, with_librtmp=False, with_libssh2=False, with_nghttp2=False, with_openssl=True, zlib:shared=False

ERROR: Missing prebuilt package for 'libcurl/7.56.1@bincrafters/stable'
Try to build it from sources with "--build libcurl"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

```
828 rmills@rmillsmm:~/gnu/github/piponazo/exiv2/build $ python /Users/piponazo/envs/conan/lib/python2.7/site-packages/conans/conan.py install .. --build libcurl
PROJECT: Installing /Users/rmills/gnu/github/piponazo/exiv2
Requirements
    Expat/2.2.5@pix4d/stable from conan-center
    gtest/1.8.0@bincrafters/stable from conan-center
    libcurl/7.56.1@bincrafters/stable from conan-center
    zlib/1.2.11@conan/stable from conan-center
Packages
    Expat/2.2.5@pix4d/stable:0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce
    gtest/1.8.0@bincrafters/stable:12dbd40f1254bd00b16aa8dece8ebae8f6b5accc
    libcurl/7.56.1@bincrafters/stable:c524ca20330d4dcedff6342b243b52f8309c04c9
    zlib/1.2.11@conan/stable:0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce

Expat/2.2.5@pix4d/stable: Already installed!
gtest/1.8.0@bincrafters/stable: Already installed!
zlib/1.2.11@conan/stable: Already installed!
libcurl/7.56.1@bincrafters/stable: WARN: Forced build from source
libcurl/7.56.1@bincrafters/stable: Building your package in /Users/rmills/.conan/data/libcurl/7.56.1/bincrafters/stable/build/c524ca20330d4dcedff6342b243b52f8309c04c9
libcurl/7.56.1@bincrafters/stable: Copying sources to build folder
libcurl/7.56.1@bincrafters/stable: Generator cmake created conanbuildinfo.cmake
libcurl/7.56.1@bincrafters/stable: Calling build()
libcurl/7.56.1@bincrafters/stable: 
libcurl/7.56.1@bincrafters/stable: ERROR: Package 'c524ca20330d4dcedff6342b243b52f8309c04c9' build failed
libcurl/7.56.1@bincrafters/stable: WARN: Build folder /Users/rmills/.conan/data/libcurl/7.56.1/bincrafters/stable/build/c524ca20330d4dcedff6342b243b52f8309c04c9
ERROR: libcurl/7.56.1@bincrafters/stable: Error in build() method, line 123
	self.build_with_autotools()
while calling 'build_with_autotools', line 321
	env_build = AutoToolsBuildEnvironment(self, win_bash=self.is_mingw)
	TypeError: __init__() got an unexpected keyword argument 'win_bash'
829 rmills@rmillsmm:~/gnu/github/piponazo/exiv2/build $ 
```

This looks confused.  It appears to want to build the mingw version of libcurl with the autotools.  Ummm.

piponazo committed:
Add additional compiler flags when external XMP is enabled on Mac

D4N commented:
@piponazo Great work! I'll try this out today or tomorrow.

I have two remarks:
1. Why not add it to the CI? If this is a viable replacement, we should test it continuously.
2. Could I hijack your repo https://github.com/piponazo/conan-xmpsdk and reuse the cmake code to package the XMP SDK for Linux distros? You have put in a lot of effort and it would be a waste of time for package maintainers to redo all of that.

piponazo commented:
IMHO I think it is not worth to dedicate many resources in supporting several versions of XMP SDK (At least not until users get interest in this new feature). For many years, people have been using the XMP version embedded in the Exiv2 repository and apparently @clanmills just got few messages from some people showing interest about being about to use newer versions of XMP.

1. Regarding your point about the CI, my opinion is the same. I would only try to check consistently the code if we start to have more users being interested on this new feature (For the moment I would say that we should consider it in experimental state).
2. Feel free to use the conan-xmpsdk as you please :). I just do not know if package maintainers are interested at all in having that library in their packaging systems.

I would try to do more work than needed related to XMP SDK. I did this work as a pre-condition to remove the autools and msvc code from the repo, so that we can only maintain one configuration/build system: the CMake one. In the next months I would like to contribute to Exiv2 and not to XMP.

clanmills commented:
Thank you for doing the work, Luis.  I know you heart isn’t it this project, however I appreciate you doing this work.  I has been requested by more than one user.  Andreas volunteered for the project about 5 years ago and I didn’t get involved.  Alien Skin have asked me numerous times to get this done.

I agree with you that this should not be our default build.  It’s an option for folks using the Adobe XMPsdk as a library in their application.  The default build should what’s been done for the last 10 years - we compile and statically link the xmpsdk/ directory into Exiv2.

I have not yet succeeded in getting a build with conan+cmake+adobe-2016 to work on Mac or Linux.  I’ll try Cygwin and msvc in the next day or two.

clanmills commented:
@piponazo:  I'm making progress with this.
1) I have successfully build on Linux
2) MSVC is failing horribly (thousands of compiler errors).  I'd like to spend more time looking at that.  Good News: contrib/cmake/msvc is building correctly, although it knows nothing about Adobe XMPsdk.
3) MacOS-X builds `-rwxr-xr-x+ 1 rmills staff 5775432 May 21 17:43 ./build/lib/libexiv2.26.0.0.dylib` and fails to link the applications such as exiv2.  I think he is trying to link gnu++98 and c++11.  For sure, he produces the warning:
```
[ 51%] Linking CXX executable ../bin/exiv2
cd /Users/rmills/gnu/github/piponazo/exiv2/build/src && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/exiv2.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -Wall -Wcast-align -Wpointer-arith -Wformat-security -Wmissing-format-attribute -Woverloaded-virtual -W -std=gnu++98 -stdlib=libstdc++ -fvisibility=hidden -fvisibility-inlines-hidden -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.11 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/exiv2.dir/exiv2.cpp.o CMakeFiles/exiv2.dir/actions.cpp.o CMakeFiles/exiv2.dir/utils.cpp.o CMakeFiles/exiv2.dir/getopt_win32.c.o  -o ../bin/exiv2 -Wl,-rpath,/Users/rmills/gnu/github/piponazo/exiv2/build/lib ../lib/libexiv2.26.0.0.dylib /Users/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a 
clang: warning: libstdc++ is deprecated; move to libc++ [-Wdeprecated]
Undefined symbols for architecture x86_64:
  "Exiv2::BasicError<char>::setMsg()", referenced from:
      Exiv2::BasicError<char>::BasicError<std::string>(Exiv2::ErrorCode, std::string const&) in exiv2.cpp.o
``` 
4) Cygwin is almost OK at building XmpSdk.  At 99% he tries to run the resource compiler and generates the error:
```
[ 99%] Building RC object XMPCore/CMakeFiles/XMPCore.dir/resource/win/XMPCore.rc.obj
cd /home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/build/ee2eb5d8ec2da33239b16bad912dce7f233bb87b/XMPCore && /bin/windres.exe -O co       ff -DBUILDING_XMPCORE_AS_STATIC=1 -DBUILDING_XMPCORE_LIB=1 -DENABLE_CPP_DOM_MODEL=1 -DHAVE_EXPAT_CONFIG_H=1 -DUNIX_ENV=1 -DXML_STATIC=1 -D       XMP_COMPONENT_INT_NAMESPACE=AdobeXMPCore_Int -DXMP_StaticBuild=1 -I/home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/build/ee2eb5d8ec       2da33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607 -I/home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/build/ee2eb5d8ec2da33239b       16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/public/include -I/home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/build/ee2eb5d8ec2da       33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/third-party/expat/public/lib -I/home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable       /build/ee2eb5d8ec2da33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/XMPCore/resource/linux -I/home/rmills/.conan/data/XmpSdk/2016.7/p       iponazo/stable/build/ee2eb5d8ec2da33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/XMPCore/resource/win   /home/rmills/.conan/data/Xmp       Sdk/2016.7/piponazo/stable/build/ee2eb5d8ec2da33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/XMPCore/resource/win/XMPCore.rc CMakeFi       les/XMPCore.dir/resource/win/XMPCore.rc.obj
/home/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/build/ee2eb5d8ec2da33239b16bad912dce7f233bb87b/XMP-Toolkit-SDK-CC201607/XMPCore/res       ource/win/XMPCore.rc:9:83: fatal error: winres.h: No such file or directory
 #include "winres.h" // ! Needed to create the Version tab in the Properties panel.
                                                                                   ^
compilation terminated.
/bin/windres: preprocessing failed.
```

The .rc file is "resource source" (strings, icons, menus and stuff like that).  The resource compiler converts .rc files into .res which get linked into executables (.exe/.dll and the like).  This file winres.rc is part of the Microsoft SDK.  I think there's something missing in the conan/CMake files to detects that file and ensure that the appropriate -Ipath-to-winres is on the command-line for the resource compiler.  Here's where those files are on the build machine:
```
522 rmills@rmillsmm-w7:~/gnu/github/piponazo/exiv2/build $ grep winres.h /c/temp/theLot.txt
        1151 .\Program Files\Microsoft SDKs\Windows\v7.0\Samples\winui\tsf\tsfapp\winres.h
        1151 .\Program Files\Microsoft SDKs\Windows\v7.0\Samples\winui\tsf\tsfcompart\winres.h
        1141 .\Program Files (x86)\Microsoft Visual Studio .NET 2003\Vc7\atlmfc\include\winres.h
        1132 .\Program Files (x86)\Microsoft Visual Studio 10.0\VC\atlmfc\include\winres.h
        1141 .\Program Files (x86)\Microsoft Visual Studio 8\VC\atlmfc\include\winres.h
        1544 .\Program Files (x86)\Microsoft Visual Studio 8\VC\ce\atlmfc\include\winres.h
        1141 .\Program Files (x86)\Microsoft Visual Studio 8 from mbp\VC\atlmfc\include\winres.h
        1544 .\Program Files (x86)\Microsoft Visual Studio 8 from mbp\VC\ce\atlmfc\include\winres.h
        1134 .\Program Files (x86)\Microsoft Visual Studio 9.0\VC\atlmfc\include\winres.h
        1537 .\Program Files (x86)\Microsoft Visual Studio 9.0\VC\ce\atlmfc\include\winres.h
        1134 .\Program Files (x86)\Microsoft Visual Studio 9.0 from mbp\VC\atlmfc\include\winres.h
        1537 .\Program Files (x86)\Microsoft Visual Studio 9.0 from mbp\VC\ce\atlmfc\include\winres.h
        1141 .\Program Files (x86)\VS2005\VC\atlmfc\include\winres.h
        1544 .\Program Files (x86)\VS2005\VC\ce\atlmfc\include\winres.h
        1207 .\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\um\winres.h
        1207 .\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\um\winres.h
        1207 .\Program Files (x86)\Windows Kits\8.0\Include\um\winres.h
        1207 .\Program Files (x86)\Windows Kits\8.1\Include\um\winres.h
523 rmills@rmillsmm-w7:~/gnu/github/piponazo/exiv2/build $
```

piponazo commented:
Hi guys. Thank you for reviewing this, I hope we can make this work together.

@clanmills I tried again the build on Mac and you are right, the situation is not completely fixed yet. When I tried by myself in your mac server I managed to compile XMPSDK with conan, and then I just tried to compile the target **exiv2lib** in the Exiv2 repository. Up to this point things compile & link correctly. But when I try to compile an app (i.e **exiv2**) I get the same linking errors you are getting:

```
clang: warning: libstdc++ is deprecated; move to libc++ [-Wdeprecated]
```
I have few more ideas to try there. Let me spend some more time on Mac, to see if I can fix those damn issues. 

Regarding to Windows, I checked again and I did not have any problem. This is the list of command I execute to compile the full project:

```
source /d/Dev/Windows/envs/conan/Scripts/activate
cd projects/exiv2/
git checkout newXmpSdk
git pull --rebase
rm -rf build/
mkdir build && cd build
conan install .. --profile msvc15Release -o xmp=True
cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_XMP=OFF -DEXIV2_ENABLE_EXTERNAL_XMP=ON -DEXIV2_ENABLE_NLS=OFF -DEXIV2_ENABLE_PNG=ON -DEXIV2_ENABLE_WEBREADY=ON -DEXIV2_ENABLE_CURL=ON -DCMAKE_INSTALL_PREFIX=install ..
ninja
```
And It also worked without ninja (Using directly the Visual Studio 2017 Generator):
```
rm -rf build && mkdir build && cd build
conan install .. --profile msvc15Release -o xmp=True
cmake -G"Visual Studio 15 2017 Win64" -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_XMP=OFF -DEXIV2_ENABLE_EXTERNAL_XMP=ON -DEXIV2_ENABLE_NLS=OFF -DEXIV2_ENABLE_PNG=ON -DEXIV2_ENABLE_WEBREADY=ON -DEXIV2_ENABLE_CURL=ON -DCMAKE_INSTALL_PREFIX=install ..
cmake --build . --config Release
```

piponazo committed:
Remove useless code for Mac

piponazo commented:
I've managed to fix the compilation on Mac. This are the command I used to compile the project:

```
mkdir build && cd build
conan install ../ -o xmp=True 
cmake -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_XMP=OFF -DEXIV2_ENABLE_EXTERNAL_XMP=ON -DCMAKE_BUILD_TYPE=Release -GXcode ../
cmake --build . --config Release
```
Note that we have to use the Xcode generator now so that it passes similar compilation flags to Exiv2 than the ones used for XmpSDK. I just committed few changes to demonstrate that now it is possible to also compile on Mac with the new XMP SDK 2016. However I am not done here, since I want to improve something on the XMP SDK recipe.

It turns out that the CMake code of XMP SDK is fixing the standard library to be used to `-stdlib=libstdc++` that is deprecated on the latest mac osx SDKs (as the compiler is telling us with some warnings). We have to switch that to `-stdlib=libc++` in both: XMP SDK and Exiv2 (When compiling with clang).

piponazo committed:
Disable 'embedded' XMP when EXIV2_ENABLE_EXTERNAL_XMP is enabled

piponazo commented:
Oh god ... the XMP SDK compilation does not work with `stdlib=libc++`. Then I will not change more things on the conan-xmpsdk recipe and I will just adapt the Exiv2 compilation flags when the `EXIV2_ENABLE_EXTERNAL_XMP` is ON.

piponazo committed:
Set libstdc++ with Xcode generator when using external XMP

D4N commented:
I tried to install the xmpsdk in a CentOS VM with conan, but unfortunately the build fails because conan tries to install the package `uuid-dev` which does not exist in CentOS or Fedora. It's called `uuid-devel` on CentOS & Fedora.

The package name seems to be hardcoded in `conanfile.py` in line 20. Could this be made variable depending on the distro? Or the installation left to the user (probably the simplest solution)?

piponazo commented:
> The package name seems to be hardcoded in conanfile.py in line 20. Could this be made variable depending on the distro? Or the installation left to the user (probably the simplest solution)?

Sure, thanks for letting me know. I will fix it this afternoon

clanmills commented:
Making progress on MacOS-X thanks to Luis.

1) Can we get the build report to report EXTERNAL_XMP for the command:

```
$ cmake .. -DEXIV2_ENABLE_EXTERNAL_XMP=On
....
-- ------------------------------------------------------------------
-- exiv2 0.26 configure results        <http://www.exiv2.org>
-- Building shared library:            YES
-- Building PNG support:               YES
-- XMP metadata support:               NO
-- Native language support:            YES
-- Conversion of Windows XP tags:      YES
-- Nikon lens database:                YES
-- Commercial build:                   NO
-- Building video support:             NO
-- Building webready support:          NO
-- Building exiv2 command:             YES
-- Building samples:                   YES
-- Building PO files:                  NO
-- Building unit tests:                NO
-- Using ccache:                       NO
-- ------------------------------------------------------------------
```

2) It's not linking the library, although it's producing sensible error messages:
```
[ 48%] Linking CXX shared library ../lib/libexiv2.dylib
cd /Users/rmills/gnu/github/piponazo/exiv2/build/src && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/exiv2lib.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Wall -Wcast-align -Wpointer-arith -Wformat-security -Wmissing-format-attribute -Woverloaded-virtual -W -std=gnu++98 -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.11 -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 26.0.0 -current_version 26.0.0 -o ../lib/libexiv2.26.0.0.dylib -install_name @rpath/libexiv2.26.dylib CMakeFiles/exiv2lib.dir/asfvideo.cpp.o CMakeFiles/exiv2lib.dir/basicio.cpp.o CMakeFiles/exiv2lib.dir/bigtiffimage.cpp.o CMakeFiles/exiv2lib.dir/bmpimage.cpp.o CMakeFiles/exiv2lib.dir/convert.cpp.o CMakeFiles/exiv2lib.dir/cr2image.cpp.o CMakeFiles/exiv2lib.dir/crwedit.cpp.o CMakeFiles/exiv2lib.dir/crwimage.cpp.o CMakeFiles/exiv2lib.dir/datasets.cpp.o CMakeFiles/exiv2lib.dir/easyaccess.cpp.o CMakeFiles/exiv2lib.dir/epsimage.cpp.o CMakeFiles/exiv2lib.dir/error.cpp.o CMakeFiles/exiv2lib.dir/exif.cpp.o CMakeFiles/exiv2lib.dir/futils.cpp.o CMakeFiles/exiv2lib.dir/gifimage.cpp.o CMakeFiles/exiv2lib.dir/http.cpp.o CMakeFiles/exiv2lib.dir/image.cpp.o CMakeFiles/exiv2lib.dir/ini.cpp.o CMakeFiles/exiv2lib.dir/iptc.cpp.o CMakeFiles/exiv2lib.dir/jp2image.cpp.o CMakeFiles/exiv2lib.dir/jpgimage.cpp.o CMakeFiles/exiv2lib.dir/matroskavideo.cpp.o CMakeFiles/exiv2lib.dir/metadatum.cpp.o CMakeFiles/exiv2lib.dir/mrwimage.cpp.o CMakeFiles/exiv2lib.dir/orfimage.cpp.o CMakeFiles/exiv2lib.dir/pgfimage.cpp.o CMakeFiles/exiv2lib.dir/preview.cpp.o CMakeFiles/exiv2lib.dir/properties.cpp.o CMakeFiles/exiv2lib.dir/psdimage.cpp.o CMakeFiles/exiv2lib.dir/quicktimevideo.cpp.o CMakeFiles/exiv2lib.dir/rafimage.cpp.o CMakeFiles/exiv2lib.dir/riffvideo.cpp.o CMakeFiles/exiv2lib.dir/rw2image.cpp.o CMakeFiles/exiv2lib.dir/tags.cpp.o CMakeFiles/exiv2lib.dir/tgaimage.cpp.o CMakeFiles/exiv2lib.dir/tiffimage.cpp.o CMakeFiles/exiv2lib.dir/types.cpp.o CMakeFiles/exiv2lib.dir/utilsvideo.cpp.o CMakeFiles/exiv2lib.dir/value.cpp.o CMakeFiles/exiv2lib.dir/version.cpp.o CMakeFiles/exiv2lib.dir/webpimage.cpp.o CMakeFiles/exiv2lib.dir/xmp.cpp.o CMakeFiles/exiv2lib.dir/xmpsidecar.cpp.o CMakeFiles/exiv2lib.dir/pngimage.cpp.o CMakeFiles/exiv2lib_int.dir/canonmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/casiomn_int.cpp.o CMakeFiles/exiv2lib_int.dir/cr2image_int.cpp.o CMakeFiles/exiv2lib_int.dir/crwimage_int.cpp.o CMakeFiles/exiv2lib_int.dir/fujimn_int.cpp.o CMakeFiles/exiv2lib_int.dir/makernote_int.cpp.o CMakeFiles/exiv2lib_int.dir/minoltamn_int.cpp.o CMakeFiles/exiv2lib_int.dir/nikonmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/olympusmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/orfimage_int.cpp.o CMakeFiles/exiv2lib_int.dir/panasonicmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/pentaxmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/rw2image_int.cpp.o CMakeFiles/exiv2lib_int.dir/samsungmn_int.cpp.o CMakeFiles/exiv2lib_int.dir/sigmamn_int.cpp.o CMakeFiles/exiv2lib_int.dir/sonymn_int.cpp.o CMakeFiles/exiv2lib_int.dir/tiffcomposite_int.cpp.o CMakeFiles/exiv2lib_int.dir/tiffimage_int.cpp.o CMakeFiles/exiv2lib_int.dir/tiffvisitor_int.cpp.o CMakeFiles/exiv2lib_int.dir/pngchunk_int.cpp.o /Users/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a /Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a /usr/local/lib/libintl.dylib /usr/lib/libiconv.dylib 
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(compress.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(crc32.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(uncompr.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(deflate.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(inflate.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(trees.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(adler32.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(inffast.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(inftrees.o)) was built for newer OSX version (10.13) than being linked (10.11)
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(zutil.o)) was built for newer OSX version (10.13) than being linked (10.11)
Undefined symbols for architecture x86_64:
  "std::string::find_first_of(char const*, unsigned long, unsigned long) const", referenced from:
      XMPUtils::RemoveProperties_v2(XMPMeta*, char const*, char const*, unsigned int) in libXMPCore.a(XMPUtils2.o)
```
3)  I've deleted the both the build directory and ~/.conan/data
Result is the same.  Doesn't link.  Same warnings about the SDK.  I suspect that's harmless.
```
ld: warning: object file (/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a(compress.o)) was built for newer OSX version (10.13) than being linked (10.11)
```
The linker messages are solid.  Looks like we're not linking the correct std c++ library.
```
Undefined symbols for architecture x86_64:
  "std::string::find_first_of(char const*, unsigned long, unsigned long) const", referenced from:
      XMPUtils::RemoveProperties_v2(XMPMeta*, char const*, char const*, unsigned int) in libXMPCore.a(XMPUtils2.o)
```

Here's the link command:
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-Wall
-Wcast-align
-Wpointer-arith
-Wformat-security
-Wmissing-format-attribute
-Woverloaded-virtual
-W
-std=gnu++98  <---- this is suspicious!!
-arch x86_64
-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk
-mmacosx-version-min=10.11
-dynamiclib
-Wl,-headerpad_max_install_names
-compatibility_version 26.0.0
-current_version 26.0.0
-o ../lib/libexiv2.26.0.0.dylib
-install_name @rpath/libexiv2.26.dylib
CMakeFiles/exiv2lib.dir/asfvideo.cpp.o
...
CMakeFiles/exiv2lib_int.dir/pngchunk_int.cpp.o
/Users/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a
/Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a
/usr/local/lib/libintl.dylib
/usr/lib/libiconv.dylib
```

4) There's a warning from CMake about the Apple SDK
```
  Manually-specified variables were not used by the project:

    XMP_OSX_SDK
```
Should we be setting this on the command line?  Or should we be setting this automatically to 10.11 when we use `EXIV2_ENABLE_EXTERNAL_XMP=On`?

5) I don't see expat being linked.  Ummm that's odd.  Perhaps it's statically linked into libXMPCore.a:
```
846 rmills@rmillsmm:~/.conan/data/XmpSdk $ otool -L ./2016.7/piponazo/stable/build/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a | grep xml
./2016.7/piponazo/stable/build/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a(xmlparse.o):
./2016.7/piponazo/stable/build/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a(xmlrole.o):
./2016.7/piponazo/stable/build/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a(xmltok.o):
847 rmills@rmillsmm:~/.conan/data/XmpSdk $ 
```

piponazo committed:
CMake: indicate if XMP external is enabled

clanmills reviewed (commented):
How you've done that is very nice.  Thanks very much.
-- XMP metadata support (EXTERNAL): YES

piponazo commented:
@clanmills Could you please tell me how are you configuring the project and triggering the compilation? For me it works consistently following this steps in your OSX server:

```
mkdir build && cd build
conan install ../ -o xmp=True 
cmake -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_EXTERNAL_XMP=ON -DCMAKE_BUILD_TYPE=Release -GXcode ../
cmake --build . --config Release
```
About the different points you commented:

1. Fixed. Now it is printing correctly if the external XMP is being used.
2. I do not get this linking error. I think it might be caused because because you are running things in a different way.
3. I do not get those warnings either. I was getting them before, but now I am passing `-stdlib=libstdc++` when using the external XMP and the Xcode generator.
4. I am not getting this warning in CMake when dealing with the Exiv2 project. That variable is only used for compiling the XmpSdk conan recipe.
5. Effectively expat is being linked statically when compiling XmpCore.

clanmills commented:
@piponazo Thanks.  Using the Xcode generator has improved matters.  I was using the default generator (Unix Makefiles, I think).  It's building and linking the library and sample applications.  There is a warning when linking addmoddel:
```
Ld build/bin/addmoddel normal x86_64
    cd /Users/rmills/gnu/github/piponazo/exiv2
    export MACOSX_DEPLOYMENT_TARGET=10.11
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -L/Users/rmills/gnu/github/piponazo/exiv2/build/bin -F/Users/rmills/gnu/github/piponazo/exiv2/build/bin -filelist /Users/rmills/gnu/github/piponazo/exiv2/build/samples/exiv2.build/Release/addmoddel.build/Objects-normal/x86_64/addmoddel.LinkFileList -Xlinker -rpath -Xlinker /Users/rmills/gnu/github/piponazo/exiv2/build/lib/Release -mmacosx-version-min=10.11 -stdlib=libstdc++ -Wl,-search_paths_first -Wl,-headerpad_max_install_names /Users/rmills/gnu/github/piponazo/exiv2/build/lib/Release/libexiv2.26.0.0.dylib /Users/rmills/.conan/data/zlib/1.2.11/conan/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libz.a /Users/rmills/.conan/data/XmpSdk/2016.7/piponazo/stable/package/0eaf3bfbc94fb6d2c8f230d052d75c6c1a57a4ce/lib/libXMPCore.a -Xlinker -dependency_info -Xlinker /Users/rmills/gnu/github/piponazo/exiv2/build/samples/exiv2.build/Release/addmoddel.build/Objects-normal/x86_64/addmoddel_dependency_info.dat -o /Users/rmills/gnu/github/piponazo/exiv2/build/bin/addmoddel
clang: warning: libstdc++ is deprecated; move to libc++ [-Wdeprecated]
```

I'm not able to run the tests on MacOS-X.  I think we'll need to run them via cmake or something (to get xcodebuild to run them).
```
888 rmills@rmillsmm:~/gnu/github/piponazo/exiv2/build $ make tests
make: *** No rule to make target `tests'.  Stop.
889 rmills@rmillsmm:~/gnu/github/piponazo/exiv2/build $ 
```

Thanks for your update about MSVC and using `--profile`.  I'll try that after lunch and before we go out.  The bedding has been planted in the garden!

clanmills commented:
@D4N and @piponazo I think Dan's right.  We can only use `$ make tests` with the "Unix Makefiles" generator.  Although I use CMake for development on Mac, I run the test suite from the command-line in the `exiv2` directory.  When I build with Xcode, I manually set EXIV2_BINDIR.  There's a script test/testMSVC.sh that sets that magic for testing msvc/ builds.

As we've agreed to adopt CMake as our only build environment, we should refactor this to ensure that we can run the test suite easily from the command-line or IDE.

piponazo commented:
@D4N I'll investigate the linking errors with the unit tests. I think I did not have that issue on my side.
@clanmills There is not much we can do about the warning:

```
clang: warning: libstdc++ is deprecated; move to libc++ [-Wdeprecated]
```
XmpSdk is using libstdc++ and we have to also use it when enabling the usage of the newer XMP SDK. 

Regarding to the Cygwin topic, I would prefer to investigate it in a later phase. There are already many things to consider in this PR for a experimental feature.

And finally, I think that we should not worry to much if `make tests` does not work with other generators but with Makefiles. Anyways, we are porting all the tests to python.

piponazo commented:
@D4N which Debian version were you using ? I cannot reproduce your issue on Ubuntu 18.04

piponazo commented:
@D4N As discussed in slack, I think this is caused because a mismatch between the default stdlib used in modern versions of gcc (`libstdc++11`) and the version used in conan by default (`libstdc++`). 

When you start using conan, it will create a default profile for you (or you can also create it with a command):

```
vagrant@ubuntu-exiv2:~/exiv2$ conan profile new --detect default
Found gcc 7
gcc>=5, using the major as version
Found clang 6.0

************************* WARNING: GCC OLD ABI COMPATIBILITY ***********************
 
Conan detected a GCC version > 5 but has adjusted the 'compiler.libcxx' setting to
'libstdc++' for backwards compatibility.
Your compiler is likely using the new CXX11 ABI by default (libstdc++11).

If you want Conan to use the new ABI, edit the default profile at:

    ~/.conan/profiles/default

adjusting 'compiler.libcxx=libstdc++11'

************************************************************************************



Profile created with detected settings: /home/vagrant/.conan/profiles/default
```
You will need to adjust 'compiler.libcxx=libstdc++11' so that the conan packages and Exiv2 use the same standard library.

D4N commented:
I can confirm that. I have modified the provisioner to automatically set `compiler.libcxx=libstdc++11` and now I can build exiv2 with unit test and the external xmp sdk.

clanmills commented:
Very good work, Gentlemen.  That's great.  Let's make sure we get this into README-CONAN.md

Incidentally, I remembered another reason to do nothing more about Cygwin/AdobeXMPsdk.  Adobe's code doesn't compile on Cygwin (with the 4.9 GCC compiler).  The changes required are a few macros to change several functions() to use Windows variants.  I was going to add some sed magic to xmpsdk/buildXMPsdk.sh to deal with that.  If somebody uses Cygwin/Adobe, they'll complain or give us a fix/patch.  We've done enough on this for now.

piponazo commented:
I just added some more information in the `README-CONAN.md` with notes about the things we have learned with this PR. Apart from that, in `WORK-IN-PROGRESS` I already added some notes about the experimental work we are doing here. 

Please let me know what else could I do to finalise the work in this PR.

clanmills commented:
@piponazo Thanks for your hard work on this.  Very much appreciated.   Curiously, we had a Redmine Issue this morning about XMP and Phil Harvey contributed by agreeing with my analysis.  And I was able to use your build with the Adobe XMPsdk in the investigation.  http://dev.exiv2.org/issues/1354

I'll review README-CONAN.md  You write good docs, so I think it's unlikely that I will suggest changes.

piponazo commented:
> Curiously, we had a Redmine Issue this morning about XMP and Phil Harvey contributed by agreeing with my analysis. And I was able to use your build with the Adobe XMPsdk in the investigation. http://dev.exiv2.org/issues/1354

What a coincidence! :stuck_out_tongue: . At least, that issue make me feel that the work being done here will be useful for some people :wink:

D4N commented on README-CONAN.md:
````diff
@@ -257,6 +255,27 @@ $ cmake ..
+compiler.libcxx=libc++       # will use -stdlib=libc++
+
+```
+Please, be aware that normally when using gcc >= 5.1, _GLIBCXX_USE_CXX11_ABI is set to 1 by default. However some linux distributions might set that definition to 0 by default. In case you get linking errors about standard c++ containers or algorithms when bringing the Exiv2 dependencies with conan, this might indicate a mismatch between the value set in **compiler.libcxx** and the default values ...
````
As far as I know no distro did that.

piponazo replied:
I took that sentence from the GCC link about the Dual ABI:

```
(Be aware that some GNU/Linux distributions configure GCC 5 differently so that the default value of the macro is 0 and users must define it to 1 to enable the new ABI.)
```
I only use Ubuntu regularly and I do not know if this happens in other popular distributions. However I noticed that some conan recipes set `_GLIBCXX_USE_CXX11_ABI` with different values depending on the **compiler.libcxx** to deal with this kind of issues. One example is available in one of the most popular conan recipes: conan-boost:

https://github.com/conan-community/conan-boost/blob/release/1.66.0/conanfile.py

D4N commented on README-CONAN.md:
````diff
@@ -257,6 +255,27 @@ $ cmake ..
+```
+Please, be aware that normally when using gcc >= 5.1, _GLIBCXX_USE_CXX11_ABI is set to 1 by default. However some linux distributions might set that definition to 0 by default. In case you get linking errors about standard c++ containers or algorithms when bringing the Exiv2 dependencies with conan, this might indicate a mismatch between the value set in **compiler.libcxx** and the default values ...
+
+As a rule of thum, set `compiler.libcxx=libstdc++11` when using a version of gcc >= 5.1. 
````
Typo: thum -> thumb

D4N commented on README-CONAN.md:
```diff
@@ -295,3 +295,12 @@ solution generated by CMake, the dependencies brought by conan will use a specif
+
+### Optional dependencies
+
+We recently added a conna option to add a modern version XMP SDK as a external dependency. This is a experimental feature, and in order to enable the user will need to call conan like this:
```
Typo: conna -> conan. 
@clanmills is it "as **a** external dependency" or "as **an** external dependency" (and the same with experimental feature)?

piponazo replied:
It should be ' an experimental ...'  I will fix it too ;)

clanmills commented:
You see, your efforts has fixed something that nobody knew about.  To be sure, this issue is cosmetic.  However there will be code in the Adobe XMPsdk that fixes something substantial.

The world is a highly connected place.  Software wires the planet in the most unexpected ways.

Thanks to you and @D4N for working on this.  I'm so happy that two such smart and nice people are contributing to Exiv2 and doing really great work.

clanmills commented:
@D4N It's **_an_** external dependency or **_an_** experimental feature.  The word **_a_** becomes **_an_** when the next word begins with a vowel (a, e, i, o, u).  So "an absolute disaster", or "an unexpected outcome".  However it's "a beautiful day" and indeed "a word of thanks".

D4N commented on WORK-IN-PROGRESS:
```diff
@@ -811,5 +813,44 @@ Modified: 2017-09-24
+    conan install .. --options xmp=True
+
+    # Configure the project with support for the external XMP version. Disable the normal XMP version
+    cmake -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_XMP=OFF -DEXIV2_ENABLE_EXTERNAL_XMP=ON -DBUILD_SHARED_LIBS=ON ..
```
Is `-DEXIV2_ENABLE_XMP=OFF` still required?

D4N reviewed (approved):
Outstanding work! I have tried it in the vagrant vms that work with conan and it worked everywhere.

piponazo commented:
@clanmills I merged this branch because it was becoming more a more difficult to rebase the changes in top of master. Please, do not hesitate to review the conan doc and create a new PR with the proposed changes. I will review it with pleasure.
